### ISSUE TITLE
More user-friendly JLD2OutputWriter

### ISSUE THREAD

**Issue by glwagner** [CliMA/Oceananigans.jl#963]
Currently the `JLD2OutputWriter` labels output in a timeseries by `model.clock.iteration`.

It might be easier to handle output if data is labeled by "save point", so it goes 0, 1, 2... This is easier to work with at the REPL, since you always know, for example, that `file["timeseries/u/1"]` exists (rather than having to manually inspect the file to figure out the iterations that data ended up being saved at).

We can add a field `iteration` so that the model iteration can be recovered, if that's needed.

**Comment by glwagner** [comment]
Changing this convention will change lines like

https://github.com/CliMA/Oceananigans.jl/blob/c572c39b291bfea49e8ca4f6cdd4cb94e8cb4893/examples/one_dimensional_diffusion.jl#L110

to

```julia
i_output = 0:length(keys(file["timeseries/t"]))
```

which may be preferred.

**Comment by ali-ramadhan** [comment]
Ah yeah I guess that's not a user-friendly feature so having the keys go 0, 1, 2, 3, ... would be easier to work with.

If you want the iteration number I think we can save it like `file["time"]` (which we already do) and `file["iteration"]`.

Hmmm, on one hand I think it should start at 1 though since Julia is 1-based indexing but 0 makes sense for initial condition...

**Comment by glwagner** [comment]
I like `0` for initial condition too. I think it's consistent with the meaning of `model.clock.time`; eg the number of iterations that have occurred.

**Comment by glwagner** [comment]
Another issue that cropped up is that some users expect to find fields at the top level of the file (rather than within a group that's obscurely labeled "timeseries". In other words, if we save `u`, we might expect the file to have

```julia
file["u/0"]
file["u/1"]
```

et cetera.

Note that the file typically has other information about the grid (`file[grid/]`), and perhaps about the model setup. So flattening the file hierarchy would mean that users cannot save output that has the same name as one of the model setup / metadata groups. This might be a good idea anyways...

@kburns your experience with dedalus could be valuable here...

**Comment by glwagner** [comment]
Just to make @ali-ramadhan 's comment a bit more clear, we should also change `t` to `time`.

**Comment by glwagner** [comment]
Not sure if https://github.com/JuliaIO/JLD2.jl/issues/238 will go produce anything but it may help with this issue.

cc @navidcy

**Comment by ali-ramadhan** [comment]
Side note: we should probably try uprading to JLD2 v0.2.0+ to make use of the nice new features although it might break a few things (https://github.com/CliMA/Oceananigans.jl/pull/888).

**Comment by kburns** [comment]
In the HDF5 outputs for dedalus, we have a "tasks" group that contains the specified output tasks, to avoid any name collisions like you mentioned.  HDF5 also lets you attach multiple scales to each dimension of the dataset, so in the "time" dimension we have scales for the output index, simulation iteration, simulation time, and real-world time.  BTW I really like that indexing from zero stuff :grin:.

**Comment by glwagner** [comment]
Here's an argument for 0 versus 1: the initial condition is special because it may not actually represent a valid model state. Eg #951. We define indexing for physical dimensions similarly in that index `0` lies outside the domain.

**Comment by glwagner** [comment]
> In the HDF5 outputs for dedalus, we have a "tasks" group that contains the specified output tasks, to avoid any name collisions like you mentioned.

I suppose `tasks` is just another name for `timeseries`. We could also call this special group `output` (which, like `tasks` for dedalus, is a bit more semantic than `timeseries`).

**Comment by glwagner** [comment]
> HDF5 also lets you attach multiple scales to each dimension of the dataset, so in the "time" dimension we have scales for the output index, simulation iteration, simulation time, and real-world time.

As discussed elsewhere, @kburns is referring to how a 2D dataset is saved as a 3D "array" (array-like object) in file. In HDF5, the representation of the data can be appended to. So data with size `(Nx, Ny)` ends up having size `(1, Nx, Ny)` when the file is first opened. After saving data a second time, the dataset has size `(2, Nx, Ny)`.

JLD2 doesn't support such an abstraction for array-like data in file. I think we get "close" to such an abstraction by using the save point as an index. We thus have

```julia
u = file["output/u"]

u["0"] # size (Nx, Ny, Nz), for example
u["0"][:, :, Nz] # size (Nx, Ny)

# and so on...
u["1"]
u["2"]
```

rather than

```julia
u[0, :, :, :] # size (Nx, Ny, Nz) if JLD2 had all of HDF5's features
```

If JuliaIO/JLD2.jl#238 goes through we could use `u[0]`, `u[1]`...

**Comment by glwagner** [comment]
I mean, we could actually provide this implementation ourself with something like

```julia
u = OutputField(file, "u")

u[i, a, b, c] # calls `getindex(u.file["output/u"][string(i)], a, b, c)`
``` 

actually with `OutputField` we could also automatically wrap data loaded from file into the appropriate field type as well.

**Comment by glwagner** [comment]
We can also nuke `dir` and `prefix` keyword arguments in favor of a single `filepath` kwarg (also discussed in #884)

**Comment by glwagner** [comment]
As discussed in #1068, another feature that's sorely needed is to "restore" model properties from file (mostly the grid...).

One way to do this is to provide a function that looks something like

```julia
grid = restore(filepath, :grid)
```

(Other names / API designs welcome.)

To make this possible (generally) I think it makes sense to add a group called `serialized` where we save the serialized properties indicated in `including`. We also might expand the `including` list.

**Comment by glwagner** [comment]
What do folks think about the name `force` for the keyword that determines whether or not to overwrite an existing file? I'm thinking that something more verbose like `force_overwrite` might be clearer.

**Comment by navidcy** [comment]
Yes. `force` is confusing... I was just looking at a script yesterday and I had to stop and thing for few seconds ... "forcing? what does forcing has to do here?..."

**Comment by glwagner** [comment]
What do you think about `force_overwrite`?

**Comment by ali-ramadhan** [comment]
What about just `overwrite`? `force_overwrite` seems kinda redundant.

**Comment by glwagner** [comment]
Less verbose!

**Comment by glwagner** [comment]
We have `FieldTimeSeries` now. I think the issue with `force` / `overwrite` needs to be resolved with #884 .
